api/query_engine.go

```diff
@@ -73,15 +95,120 @@ func getOptions(aggSettings mdata.AggSettings, minInterval, tsRange uint32) ([]a
 	return options, aggRef
 }
 
+// getOptionsLikeGraphite is like getOptionsLegacy but returns a sorted (by interval, asc) slice of narchives
+// in which we use
+// the actual interval we would need for raw (which is the LCM of the different raw intervals),
+// potentially making the "raw interval" > first rollup interval, so that
+// raw archive might be ordered after a rollup, and where the archive
+// itself knows which one it is, instead of having to rely on index lookup table.
+func getOptionsLikeGraphite(reqs []models.Req, aggSettings mdata.AggSettings, tsRange uint32) ([]narchive, map[uint32]struct{}) {
+	_, rawIntervals := summarizeRawIntervals(reqs)
+
+	keys := make([]uint32, len(rawIntervals))
+	i := 0
+	for k := range rawIntervals {
+		keys[i] = k
+		i++
+	}
+	lcm := util.Lcm(keys)
+
+	options := make([]narchive, 0, len(aggSettings.Aggs)+1)
+	options = append(options, narchive{0, lcm, tsRange / lcm, aggSettings.RawTTL, false})
+
+	// now model the archives we get from the aggregations
+	// note that during the processing, we skip non-ready aggregations for simplicity
+	for i, agg := range aggSettings.Aggs {
+		if agg.Ready {
+			options = append(options, narchive{i + 1, agg.Span, tsRange / agg.Span, agg.TTL, false})
+		}
+	}
+
+	sort.Sort(narchives(options))
+	return options, rawIntervals
+}
+
+func alignRequests(reqs []models.Req, aggSettings mdata.AggSettings, likeGraphite bool) ([]models.Req, error) {
+	if likeGraphite {
+		return alignRequestsLikeGraphite(reqs, aggSettings)
+	}
+	return alignRequestsLegacy(reqs, aggSettings)
+}
+
+// alignRequestsLikeGraphite is the new approach which selects the highest resolution possible, like graphite does.
+// note: instead of the traditional approach where we have the raw archive with min interval, and then compensate later
+// this mechanism tackles this aspect head-on.
+func alignRequestsLikeGraphite(reqs []models.Req, aggSettings mdata.AggSettings) ([]models.Req, error) {
+
+	tsRange := (reqs[0].To - reqs[0].From)
+	options, rawIntervals := getOptionsLikeGraphite(reqs, aggSettings, tsRange)
+
+	// find the highest res archive that retains all the data we need.
+	// fallback to lowest res option (which *should* have the longest TTL)
+	var selected int // index of selected archive within options
+	now := uint32(time.Now().Unix())
+	for i, opt := range options {
+		selected = i
+		if now-opt.ttl <= reqs[0].From {
+			break
+		}
+	}
+
+	if LogLevel < 2 {
+		options[selected].chosen = true
```

Comment:
Even though this saves an operation in prod, I wouldn't modify such an important data structure under a condition that depends on the loglevel even if at the moment it isn't used at any other critical location (not considering `narchive.String()` critical) because I think it's too easy for somebody to use that `.chosen` flag for something else without being aware that it's correctness depends on such an arbitrary setting like the log level.